### PR TITLE
Refactor the questionid ordering function

### PR DESCRIPTION
**Pull request by FelixDiLenarda**
Ordering the ids in one atomic sql querry now. It should increase the performance and it allows the deletion of `block_coursefeedback_execute_sql_arr()`

Checkonly mode was not used and was removed.

Since we only use the function when it is needed (when  questions are deleted), we don't need to check again for the need of ordering.

Also no use was made of the returns of the function. The return is now simplified to normal moodle standards.

Please note: This pull request is part of resolving https://github.com/eberhardt/moodle-block_coursefeedback/issues/47. All possible functions from lib.php will be moved to dedicated classes in a later step.

### CODE REVIEW

File: lib.php [hunk 1]
```diff
@@ -29,52 +29,43 @@
...
- * @param int $feedbackid
- * @param bool $checkonly don"t change database entries
- * @return 0 if operation failed or order is incorrect (checkonly), 1 if order is correct and 2 if order has succesful been changed
+ * @param int $feedbackid The feedback instance ID.
+ * @return bool true
+ * @throws dml_exception A DML specific exception is thrown for any errors.
  */
-function block_coursefeedback_order_questions($feedbackid, $checkonly = true) {
+function block_coursefeedback_order_questions(int $feedbackid): bool
+{
```
Comment (daniil-berg) [inline lib.php]: Not to be pedantic, but the [Moodle Coding Style guide](https://moodledev.io/general/development/policies/codingstyle#function-and-method-declaration) says:

> As with classes, the brace should always be written on same line as the function name.

File: lib.php [hunk 2]
```diff
@@ -29,52 +29,43 @@
...
-            return 1;
-        } else if (!$checkonly) {
-            if (block_coursefeedback_execute_sql_arr($sql)) {
-                return 2;
-            } else {
-                return 0;
-            }
-        }
-    }
+    // "mapping" generates a new consecutive ID (new_qid) for each distinct questionid by using 'row_number()'
```
Comment (daniil-berg) [inline lib.php]: All other variables/placeholders/aliases are nicely spelled out (albeit missing word-separators for Moodle-reasons), which I think is great because abbreviations almost always hurt readability. I would personally prefer `new_qid` to receive the same treatment, i.e. naming it `new_questionid` (or even `newquestionid`, although that hurts a bit).

File: lib.php [hunk 3]
```diff
@@ -784,40 +775,6 @@ function block_coursefeedback_create_activate_button($feedbackid, $value = "") {
...
 
-/**
- * Reimplementation of the moodle 1.9 execute_sql_arr.
- *
- * Secrurity WARNING: All statements won't be validated, before they are executed!
- *
- * @param array $sqlarr Each field is one query, one query contains the query string (key "query") and his parameters (key "params")
- * @return boolean
- */
-function block_coursefeedback_execute_sql_arr($sqlarr) {
```
Comment (daniil-berg) [inline lib.php:795]: This function is still being called from `block_coursefeedback_swap_questions` here:

https://github.com/eberhardt/moodle-block_coursefeedback/blob/1800cd127e22f3ce359a9d91317b23bf89fc7d1a/lib.php#L276

**Review by daniil-berg — CHANGES_REQUESTED** [review]
I like the idea of using a single query for this.

File: lib.php (hunk 3 above)
Comment (FelixDiLenarda) [reply lib.php:795]: Refactored the function. The unique constraint (feedbackid, questionid, language) made it somewhat comlicated.

Furthermore the logic changed. The old function swaped to questions. But in the ui says move (verschieben).
Therefore i changed it to an actual movelogic....

**Comment by FelixDiLenarda** [comment]
I think i need to refactor the ordering function again, also there in some cases the unique constraint (feedbackid, questionid, language) is violatet :(

**Comment by daniil-berg** [comment]
> I think i need to refactor the ordering function again, also there in some cases the unique constraint (feedbackid, questionid, language) is violatet :(

I am not sure understand what you mean about the unique constraint being violated.

But I was about to suggest that this is a really good candidate for thorough unit testing. Maybe it actually makes sense to go "test-first" in this instance. Write a few test cases and try to come up with different edge cases. Then see, if you can get the function to perform as expected.

Aside from the fact that you will already have tests for that function (which is good by itself obviously), such an excercise could provide additional insights. What do you think?

**Comment by FelixDiLenarda** [comment]
> > I think i need to refactor the ordering function again, also there in some cases the unique constraint (feedbackid, questionid, language) is violatet :(
> 
> I am not sure understand what you mean about the unique constraint being violated.
> 
> But I was about to suggest that this is a really good candidate for thorough unit testing. Maybe it actually makes sense to go "test-first" in this instance. Write a few test cases and try to come up with different edge cases. Then see, if you can get the function to perform as expected.
> 
> Aside from the fact that you will already have tests for that function (which is good by itself obviously), such an excercise could provide additional insights. What do you think?

The `block_coursefeedback_questns` Table has 
`        <INDEX NAME="block_cfb_qidfidlng_idx" UNIQUE="true" FIELDS="questionid, coursefeedbackid, language"/>`

Sometimes it happened (mostly when several languages existed) that this constraint was violated. I fixed it by also moving the questionids to an offset. (Like in `block_coursefeedback_move_question()` )

Yes tests would be good but i'd prefer to move the functions first into their classes since they are probably renamed then...

**Comment by daniil-berg** [comment]
Ah, I understand what you mean now. I suppose the offset approach should work well enough. I also for some reason assumed the DB engine would be "smart enough" to only check the constraints at the end of the `UPDATE` operation rather than after each affected row. Good to know that is something to look out for!

File: lib.php [hunk 4]
```diff
@@ -216,67 +229,91 @@ function block_coursefeedback_insert_question($question, $feedbackid, $questioni
...
+            $sql = "UPDATE {block_coursefeedback_questns}
+                       SET questionid = questionid - (1 + :offset)
+                     WHERE coursefeedbackid = :feedbackid
+                           AND questionid BETWEEN (:oldposoffset + 1) AND :newposoffset";
+            $DB->execute($sql, [
+                'newposoffset' => $newposition + $offset,
+                'oldposoffset' => $oldposition + $offset,
+                'offset'       => $offset,
+                'feedbackid'   => $feedbackid,
+            ]);
```
Comment (daniil-berg) [inline lib.php]: Since you repeat the same call with the same array argument below, you could move the `$DB->execute(...)` call out of and under the condition blocks.

File: lib.php [hunk 5]
```diff
@@ -29,50 +29,63 @@
...
+                        FROM mapping
+                       WHERE t.coursefeedbackid = :feedbackid2
+                             AND t.questionid = mapping.questionid";
+
+        $params = [
+            'feedbackid1' => $feedbackid,
+            'feedbackid2' => $feedbackid,
+        ];
+
+        $result = $DB->execute($sqlupdate, $params);
```
Comment (daniil-berg) [inline lib.php:81]: Not a big deal, but the `$result` variable is unnecessary since `$DB->execute(...)` always returns `true`. I keep forgetting such things and this assignment and later return of the variable would confuse me or rather suggest that the value matters, even though it does not.

As an aside, I would suggest changing the return type of this and similar functions to `void` for basically the same reason. Returning one fixed value serves no purpose, whereas the return type `void` clearly communicates to the user that the function just "affects things" (i.e. side effects), but returns nothing.

But maybe there is a good reason that I just did not think of.

File: lib.php [hunk 6]
```diff
@@ -216,67 +229,91 @@ function block_coursefeedback_insert_question($question, $feedbackid, $questioni
...
+            'newposition'  => $newposition,
+            'feedbackid'   => $feedbackid,
+        ]);
+    } catch (Exception $e) {
+        $transaction->rollback($e);
+        throw $e;
     }
+    $transaction->allow_commit();
+
+    return true;
```
Comment (daniil-berg) [inline lib.php:310]: See above.

File: lib.php [hunk 7]
```diff
@@ -216,67 +229,91 @@ function block_coursefeedback_insert_question($question, $feedbackid, $questioni
...
 /**
- * @param int $feedbackid
- * @param int $oldpos
- * @param int $newpos
- * @return bool Success of operation
+ * Moves a question within the course feedback block to a new position.
+ *
+ * This function reorders questions by updating their questionid values.
+ * It uses a temporary offset to avoid unique constraint conflicts and performs
+ * the update in a transaction.
```
Comment (daniil-berg) [inline lib.php]: Again, I am being very pedantic now... This is very valuable information, _if_ someone is interested in the _implementation_ of this function. But such details should accompany the relevant lines of code in the form of comments IMO, not be presented to the caller/user in the documentation. Like you did with `block_coursefeedback_order_questions`. But this is nothing I would request changes over.

**Review by daniil-berg — COMMENTED** [review]
LGTM in the important parts. I just have a few small suggestions.

File: lib.php (hunk 7 above)
Comment (FelixDiLenarda) [reply lib.php:237]: Generally i agree. However I'm opting to keep returning true. Explicitly returning true seems to be a common pattern in Moodle—it clearly indicates success and keeps the code consistent with other similar functions.
Furthermore it makes automated tests a little easier i think.

File: lib.php (hunk 6 above)
Comment (FelixDiLenarda) [reply lib.php:310]: see above

File: lib.php (hunk 5 above)
Comment (FelixDiLenarda) [reply lib.php:81]: Generally i agree. However I'm opting to keep returning true. Explicitly returning true seems to be a common pattern in Moodle—it clearly indicates success and keeps the code consistent with other similar functions.
Furthermore it makes automated tests a little easier i think.

File: lib.php (hunk 4 above)
Comment (FelixDiLenarda) [reply lib.php]: Solved in [df35457](https://github.com/eberhardt/moodle-block_coursefeedback/pull/56/commits/df354570c52fece0f1af4755cb4b54ada2a7f682)

File: lib.php (hunk 7 above)
Comment (FelixDiLenarda) [reply lib.php]: Resolved in [6a736cc](https://github.com/eberhardt/moodle-block_coursefeedback/pull/56/commits/6a736cc5861ccccef1dae1819260ede214348843)

File: lib.php (hunk 5 above)
Comment (daniil-berg) [reply lib.php:81]: I am not insisting. I enjoy these discussions. So, just for the sake of argument:

"X is a common pattern in Y-codebase" is not an argument by itself. X might be bad pattern, regardless of the authority of Y. But if your goal is to emulate Moodle code style, even above what they explicitly state in their docs, I guess this makes sense.

The function not throwing an error communicates exactly the same amount of information to the caller as returning `true` (or `1` or `'foo'`). Conversely, if an error is thrown, the function obviously does not return at all. In other words, when the caller sees the function returning `void`, this is what he would reasonably expect ("no exception = all is fine"); there is nothing left to guess. When the caller sees the return type `bool`, he may reasonably expect the function to return `false` on failure and even write checks/conditions for that return value, unless he also reads the return type annotation _in the PHPDoc_ telling him `true` is the only possible value. Other languages do allow return type annotations of literals, such as `Literal[True]` in [Python for example](https://docs.python.org/3/library/typing.html#typing.Literal), but PHP unfortunately does not.

As for how it makes testing easier, I am genuinely curious. Can you give an example?

**Review by daniil-berg — APPROVED** [review]

File: lib.php (hunk 5 above)
Comment (eberhardt) [reply lib.php:81]: I would agree with Felix on this topic. Returning true might be uneseccary, but if used in an if returning void would mean false. Therefore, even if the function is successful giving the suggestion something went wrong.
That it always returning true is also giving false impressions is true, but I think it get outweight by the previous thoughts. 

In regard to address the issue, make it clear in the doc string, that this function always returns true and errors need to be adressed by try-catch.

Comment (eberhardt) [reply lib.php:81]: Though on further thought I lean toward Daniel again... it's a haze.

Comment (daniil-berg) [reply lib.php:81]: > [...] Returning true might be uneseccary, but if used in an if returning void would mean false. Therefore, even if the function is successful giving the suggestion something went wrong. [...]

First I wanted to reply that you cannot use a `void` return, but then I remembered we are dealing with PHP here... :laughing: I forgot that the `void` return type annotation is a pointless lie because the [PHP documentation](https://www.php.net/manual/en/language.types.void.php) literally says so:

> Even if a function has a return type of [void](https://www.php.net/manual/en/language.types.void.php) it will still return a value, this value is always [null](https://www.php.net/manual/en/reserved.constants.php#constant.null).

Why on earth does this even exist then?! They could just as easily allow annotating the function return with `null` then. That would even be more honest. Like Python does with `None` as the return type annotation.

Anyway, you are right, `null` is of course _falsy_. And I understand the argument. If the caller is _used_ to functions distinguishing success from failure with a `bool` return, then returning `null` could cause a problem.

I just think that functions _should not do that_. A `bool` just never carries enough information to be valuable for error handling. Something like Rust has the [`Result`](https://doc.rust-lang.org/std/result/) enum, which is specifically designed for this purpose. In PHP we are stuck with exceptions. Still, PHP seems to enjoy distinguishing success from failure by `bool` returns in many built-in functions.

But to bring us back to the actual functions in question here:

What would a caller expect the return value of a function called `order_questions` or `move_question` to tell him? Why would he use it in an `if`-condition?

```php
if (move_question(...)) {
    // Yes, question was moved.
} else {
    // I think something went wrong... But I will never know what.
    // Also, this block is literally unreachable!
}
```

I would opt for being transparent and to expect the _bare minimum_ of responsibility of the caller: Check the return type before you use the function.

Side note: I would find this _less_ objectionable, if we were on [PHP 8.2](https://www.php.net/releases/8.2/en.php#null_false_true_types) or higher with this codebase because then we could just use the `true` literal itself in the return type annotation. I still think it is pointless, but at least it would be unambiguous.

It is obviously your call, @eberhardt.

Comment (FelixDiLenarda) [reply lib.php:81]: I totally get your points, Daniel, and I would probably do it your way when writing a new codebase.
In Moodle, I think it's consistent, and I don't think it's really harmful.
